src/loki_coordinator.py

```diff
@@ -0,0 +1,241 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical
+# See LICENSE file for licensing details.
+
+"""Loki coordinator."""
+
+import logging
+from collections import Counter
+from pathlib import Path
+from typing import Any, Dict, Iterable, Optional, Set
+
+from loki_cluster import (
+    LOKI_CERT_FILE,
+    LOKI_CLIENT_CA_FILE,
+    LOKI_KEY_FILE,
+    LokiClusterProvider,
+    LokiRole,
+)
+from loki_config import _S3ConfigData
+
+logger = logging.getLogger(__name__)
+
+MINIMAL_DEPLOYMENT = {
+    LokiRole.read: 1,
+    LokiRole.write: 1,
+    LokiRole.backend: 1,
+}
+"""The minimal set of roles that need to be allocated for the
+deployment to be considered consistent (otherwise we set blocked). On top of what loki itself lists as required,
+we add alertmanager."""
+
+RECOMMENDED_DEPLOYMENT = Counter(
+    {
+        LokiRole.read: 1,
+        LokiRole.write: 1,
+        LokiRole.backend: 1,
+    }
+)
+"""The set of roles that need to be allocated for the
+deployment to be considered robust according to the official recommendations/guidelines."""
+
+# The minimum number of workers per role to enable replication
+REPLICATION_MIN_WORKERS = 3
+# The default amount of replicas to set when there are enough workers per role;
+# otherwise, replicas will be "disabled" by setting the amount to 1
+DEFAULT_REPLICATION = 3
+
+
+class LokiCoordinator:
+    """Loki coordinator."""
+
+    def __init__(
+        self,
+        cluster_provider: LokiClusterProvider,
+        # TODO: use and import tls requirer obj
+        tls_requirer: Any = None,
+        # TODO: use and import s3 requirer obj
+        s3_requirer: Any = None,
+        # root and recovery data need to be in distinct directories
+        root_data_dir: Path = Path("/data"),
+        recovery_data_dir: Path = Path("/recovery-data"),
+    ):
+        self._cluster_provider = cluster_provider
+        self._s3_requirer = s3_requirer  # type: ignore
+        self._tls_requirer = tls_requirer  # type: ignore
+        self._root_data_dir = root_data_dir
+        self._recovery_data_dir = recovery_data_dir
+
+    def is_coherent(self) -> bool:
+        """Return True if the roles list makes up a coherent loki deployment."""
+        roles: Iterable[LokiRole] = self._cluster_provider.gather_roles().keys()
+        return set(roles).issuperset(MINIMAL_DEPLOYMENT)
```

Comment:
Does this look clearer?
```suggestion
        roles: Iterable[LokiRole] = self._cluster_provider.gather_roles()
        return all(r in roles for r in MINIMAL_DEPLOYMENT)
```
or even
```suggestion
        return not self.missing_roles()
```

Reply:
`set(roles).issuperset(MINIMAL_DEPLOYMENT)` is a lot more immediate to me because it reads like English, but we can change it if you think it's better :)